### ISSUE TITLE
compiler/runtime: logical plan → equivalent physical plan without changing authority

### ISSUE THREAD

**Issue by ThibautMelen** [supernovae-st/nika#1668]
## 2026-09-19 — P2 physical optimization does not block product proof

Retain this runtime-owned equivalent-physical-plan scope after the first authoring loop. Logical source, requested authority, human-gate dominance and effect uncertainty remain fixed. No hidden parallelism/retry/routing that changes admitted behavior. Require baseline, equivalence witnesses and phase latency/cost receipts before promoting an optimization. No new public IR or alternate assembler.

## Trackable work

**P2.** Runtime / admission-plane optimizer. Parent [#1636](https://github.com/supernovae-st/nika/issues/1636) (sibling of routing [#1637](https://github.com/supernovae-st/nika/issues/1637)). Consumes **already admitted** logical programs from [#1663](https://github.com/supernovae-st/nika/issues/1663). Authoring compiler does **not** emit a physical plan. Authority invariant owned by [#1637](https://github.com/supernovae-st/nika/issues/1637) / [#1636](https://github.com/supernovae-st/nika/issues/1636): physical HTTP ⊆ admitted seats. Wire that *can* batch: [#1655](https://github.com/supernovae-st/nika/issues/1655) `questions_per_call=N`. Measurement: [#1656](https://github.com/supernovae-st/nika/issues/1656) BATCHING / PHYSICAL PLAN lane. Parameter contract honesty: [#1638](https://github.com/supernovae-st/nika/issues/1638).

### Why this is not #1637 or #1638

| Issue | Owns |
|---|---|
| #1637 | **which** seats exist in the admitted set; boring selector; hostile prompt cannot enlarge the set; `COMPILE SMART, ROUTE BORING` as *law* |
| #1638 | requested → normalized → wire → effective for **one** call's knobs; refuse-not-drop |
| #1655 | System One adapter; one HTTP; `questions_per_call` *shape* |
| **this** | **equivalence-preserving scheduling** of an *already admitted* logical DAG: batch independent decisions, parallelize independent nodes, fuse provider calls, CSE pure lookups, place seats **inside** the admitted set |

#1637 answers « may we contact TypeSafe at all? ». This issue answers « given four independent admitted DECIDE nodes on shared state, may we use one HTTP without lying in Proof? ». Merging them would bury the optimizer under the routing law. #1638 does not own multi-call fusion.

### Empirical starting point (2026-09-17)

- [x] TypeSafe docs: questions evaluated **in parallel and in isolation**; adding questions barely changes latency; *atomic questions composed in code*. Vendor claim — re-measure on #1656.
- [x] #1655 v1 already compiles N closed properties on one object into **one** HTTP. Hierarchical/dependent stays sequential `with:` tasks. Adapter must not hide `MAX_DEPTH=12`.
- [x] CheckReport already projects `access_plan` · `cost.tasks[].attempts` · `certificate`. No `PhysicalPlan` struct. Do **not** invent one until a first transformation exists and ExecutionSnapshot is audited.
- [x] `classify-and-route` is one DECIDE (`extract_facts`) not four. First fusion fixture must be **authored** as four independent closed-enum infers sharing `with:` state — do not flatten the golden to make batching look good.

### Hard invariant (non-negotiable)

Physical optimization may **NOT** change:

```
policy
permits
authority
required outputs
effect semantics
human gates
logical Proof answers (four decisions stay four answers)
admitted seat set          (#1637)
requested→effective knobs  (#1638)
```

```
logical DAG  ≠  transport call graph
physical HTTP  ⊆  access_plan
local-only + TYPESAFE_API_KEY set  →  0 TypeSafe HTTP     (#1637)
compiler / optimizer authority = NONE
```

### Ship (ONE transformation)

- [ ] Audit ExecutionSnapshot / Proof / DAP: where would four batched answers live today without a new IR? If they cannot, **stop** and extend Proof projection — do not invent `PhysicalPlan`.
- [ ] Fixture: 4 independent closed-enum `infer` tasks, identical `with:` state, no `when:` coupling, no effects. PHYSICAL-A: 4 HTTP. PHYSICAL-B: 1 System One (or 1 compatible) request with 4 questions.
- [ ] Proof / trace names **each** logical answer, requested alias, returned model, per-question selected/probs. If PHYSICAL-B collapses to one blob, that cell is a **#1656 BATCHING loss**, not a win of this issue.
- [ ] Failure semantics: one question 422/timeout must not be reported as four successes. Document isolation. If the vendor evaluates in isolation but the adapter cannot project a per-question error, refuse to batch.
- [ ] Check still sees 4 logical tasks (cost floor, attempts). Fusion is a physical choice **inside** the admitted envelope, not a rewrite of the file the author committed — **or** the file stays 4 tasks and the engine batches (preferred: engine-side, so the logical program remains the source).
- [ ] `local-only` fixture with TypeSafe key present: 0 TypeSafe HTTP (cross-issue with #1637; this issue must not regress it).
- [ ] Measure on #1656 PHYSICAL PLAN lane: latency, billing_units vs USD vs unknown, failure isolation, Proof size, observability. Do not import TypeSafe 70–500 ms.
- [ ] `proven_by:` cassette job for PHYSICAL-B shape + isolation failure fixture.

### Batch eligibility (must all hold, else PHYSICAL-A)

same provider/seat · same admitted model · same shared state identity · questions independent · no A→B dependency · compatible timeout · compatible custody · compatible retry policy · compatible failure projection · **no effect between questions**

- [ ] **Equivalence witness:** logical answer vector A ≡ vector B under cassette. Preserve task identity, policy inputs, authority, effect ordering. Failure injection required
- [ ] **Identity:** engine-side batching of the *same* logical file does **not** change [#1664](https://github.com/supernovae-st/nika/issues/1664). Source rewrite does. Proof records physical treatment
- [ ] **Param contract:** fused call still projects requested/normalized/wire/effective **per logical question** (#1638). No silent timeout/schema/thinking drop
- [ ] Intra-task N properties in **one** infer = [#1655](https://github.com/supernovae-st/nika/issues/1655). Cross-task fusion = **this** issue. Do not merge

### ROADMAP (not this close)

- [ ] Parallelize independent non-decision nodes (already mostly DAG waves — measure before inventing).
- [ ] Common lookup elimination when pure.
- [ ] Cache pure COMPUTE with identity (#1664) — only if purity is Check-visible.
- [ ] Authorized seat *placement* (local vs cloud) as a physical choice **inside** `access_plan`, never a way to add a seat.
- [ ] FlowCompile class (arXiv:2605.13647): compile-time design-space of seats/budgets → reusable authorized configurations. **Not** runtime improvisation. Selector stays boring (#1637). Do not import 6.4×.

## Socratic

If four judgments share a state, why pay four round-trips — and if we fuse them, what happens to the Proof of the one that failed?

## Kill / reject

- Batching damages failure isolation or Proof per-decision observability → **reject the treatment**, keep PHYSICAL-A.
- Fusion changes who is contacted vs `access_plan` → security bug, not an optimization.
- Silent retry-into-unbatched on 429 that alters spend → #1637/#1573 territory; this issue must not add it.
- Flattening the **logical** YAML to one infer to make the bench look cheap → failed scientific integrity.

## Negative

`physical_plan:` YAML key. Second public IR. One SDK `PhysicalPlan` L0. Runtime « maybe batch, maybe not » depending on a prompt. Hiding four answers behind one selected label. Growing #1637's close gate with this work (V1 routing still ships without fusion).

## Alternatives considered

- **Put this in #1637 later slice** — rejected: routing law and optimizer passes have different close proofs and different failure modes.
- **Put this in #1655** — rejected: adapter owns one-call compile of *one task's* schema; this owns *multi-task* equivalence.
- **Put this in #1656** — rejected: #1656 measures; this implements.

## Acceptance (v0 = batching only)

1. 4-decision fixture: PHYSICAL-A and PHYSICAL-B produce the same four logical answers on cassette.
2. Proof names four answers in B.
3. Injected per-question failure is visible, not four greens.
4. `access_plan` unchanged. `local-only` still 0 TypeSafe HTTP.
5. No new envelope key. ExecutionSnapshot audit note in the PR.
6. #1656 lane can show B *losing* on observability.

## Dependencies

Consumes logical programs from #1663 · parent remains [#1636](https://github.com/supernovae-st/nika/issues/1636) · invariant #1637 · knobs #1638 · wire #1655 · measure #1656 · identity of fused vs unfused is **not** a program change if the logical file is unchanged (engine-side batch). If we rewrite the file to one infer, that **is** a program change (#1664) — prefer engine-side.

## Proof to close

Cassette A vs B + isolation fixture. `proven_by:` job. Observability loss published, not buried.

Siblings: prepared [#1665](https://github.com/supernovae-st/nika/issues/1665) · algebra [#1666](https://github.com/supernovae-st/nika/issues/1666) · lowering [#1667](https://github.com/supernovae-st/nika/issues/1667). Invariant parent-law: [#1637](https://github.com/supernovae-st/nika/issues/1637).

### Analogy (engineering, not marketing)

SQL query describes *what*; optimizer describes *how*. Workflow = meaning. Physical plan = batch / parallel / place. **Never** change policy, authority, outputs, effects.

- [ ] 4-decision cassette (category, urgency, churn, fraud). Proof still names four answers.

### Invisible to `compile()`

SDK/Serve authoring returns the **logical** workflow. Physical treatment is not an authoring format. Runtime Proof may name it at advanced depth ([#1669](https://github.com/supernovae-st/nika/issues/1669)). [nika-client#128](https://github.com/supernovae-st/nika-client/issues/128) must not grow `physicalPlan`.

Must **not** shape Compile V0 source architecture. Foundational compiler emits the **logical** program. Physical batching stays later.

TypeSafe “parallel and isolated” ≠ statistically independent evidence. Physical batch of four DECIDE nodes must preserve **independence_group** so [#1657](https://github.com/supernovae-st/nika/issues/1657) cannot treat four same-call answers as four sources. Proof must still name four answers. Observability loss = treatment loses ([#1656](https://github.com/supernovae-st/nika/issues/1656) BATCH).

Not a child of [#1663](https://github.com/supernovae-st/nika/issues/1663). FlowCompile ([arXiv:2605.13647](https://arxiv.org/abs/2605.13647)) is later measurement — **do not import 6.4×**.

Logical QuestionPlan (N independent compile questions) may become **one** physical infer iff they share state and independence — same law as runtime physical batching. Provenance keeps N logical decisions. Owner of the plan is [#1663](https://github.com/supernovae-st/nika/issues/1663); this issue keeps physical equivalence.

### Physical fusion is post-admission only

Logical program → **Check/admission** ([#1636](https://github.com/supernovae-st/nika/issues/1636)) → this optimizer.

May batch independent DecisionEvidence questions, parallelize independent pure reads, CSE pure results **iff** semantics and evidence remain equivalent. Must not hide 4 logical decisions as 1 Proof atom. Compile must not emit transport optimization.

## Verified work stdlib research addendum — 2026-09-18

**Status: RESEARCH / PROPOSAL. Existing bounded v0 close criteria are preserved.**

### Classify reliability optimizations on the physical plane

Preserve the existing batching v0 close gate. Later safe-read hedging/speculation and operational admission control need separate equivalence/cost evidence; they are not new semantic patterns by default.

- [ ] HEDGED_READ requires genuinely safe operations, bounded copies/delay/deadline, throttling/cancellation and first-success versus first-valid distinction. Cancellation is not proof an effect did not occur. Do not hedge mutations.
- [ ] Compare single request and hedged reads under the same workload; report tail latency, added load/cost and failure amplification. Retire hedging where it worsens the declared trade-off.
- [ ] QUEUE_BOUND/BACKPRESSURE/LOAD_SHED are runtime policy unless overload disposition is an explicit user semantic requirement. Preserve visible refusal/deadline outcome and do not silently drop work.
- [ ] ADVISOR→EXECUTOR or more agents require equal-budget baseline evidence (#1656); provider selection cannot alter admitted policy/custody/tool scope.

Do not introduce windows for finite batch work or equate scheduling watermarks with event-time completeness. Kill optimization that changes ordering/effect/authority/failure meaning or hides costs/loser outcomes.

### Nika empirical baseline

Audited engine `9d554c84c8a63144c36e7245fee641e6bfc7349f` and Spec `4b6eaadde483bcc9db9c05b022afbedfb107f37e` on 2026-09-18. [Compile](https://github.com/supernovae-st/nika/blob/9d554c84c8a63144c36e7245fee641e6bfc7349f/crates/nika-onboard/src/compile/mod.rs) is an exact-skeleton/conservative-edit foundation; [private facets](https://github.com/supernovae-st/nika/blob/9d554c84c8a63144c36e7245fee641e6bfc7349f/crates/nika-onboard/src/compile/pattern.rs) are version 0. [The development corpus](https://github.com/supernovae-st/nika-spec/tree/4b6eaadde483bcc9db9c05b022afbedfb107f37e/eval/hot) contains 22 canonical references, 220 mapped families, 0 PROMOTED_HOT and 0 evaluated compiler-generated candidates. These observations are not qualification of the proposed feature.

### Scientific evidence and transfer limits

- **PRODUCTION_DOCS** — [gRPC request hedging](https://grpc.io/docs/guides/request-hedging/). Transfer: Bound duplicate requests, delay, throttling, winner and cancellation.. Does not transfer: Hedged mutations are rejected by default; cancellation cannot undo a request.. Empirical basis: Official client behavior, no universal latency win..
- **PRODUCTION_DOCS** — [Using load shedding to avoid overload, Amazon Builders' Library](https://aws.amazon.com/builders-library/using-load-shedding-to-avoid-overload/). Transfer: Bound admission queues and reject before expensive work.. Does not transfer: Operational scheduling policy is not automatically user workflow semantics.. Empirical basis: Engineering experience and guidance; no Nika throughput claim..
- **PREPRINT** — [Scaling Test-time Compute for LLM Agents, v1 (2025)](https://arxiv.org/abs/2506.12928v1). Transfer: Compare sampling, revision, verifier/merge and rollout diversity at equal budget.. Does not transfer: No automatic default BEST_OF_N; reported ranking may depend on tasks/verifiers.. Empirical basis: Ablation study, not reproduced here..
- **NIKA_MEASUREMENT** — [Pinned Spec inventory and local integrity checks](https://github.com/supernovae-st/nika-spec/tree/4b6eaadde483bcc9db9c05b022afbedfb107f37e/eval/hot). Transfer: Separate inventory, static oracle, offline execution and generated-candidate evidence.. Does not transfer: A valid hand-authored fixture or a green validator is not HOT promotion.. Empirical basis: This task: 22 digest-verified registry rows; 220-family recount; 16 inventory selftests and 49 judge selftests; behavior harness 1 passed/7 skipped..
